pkg/hash/configmap.go

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ */
+
+package hash
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/pkg/errors"
+)
+
+// hash package purpose is to compute a ConfigMap hash
+// that will be attached as an annotation to workload resources (DaemonSet, Deployment, etc.)
+// in order to allow them to track ConfigMap changes
+// more about this technique here: https://blog.questionable.services/article/kubernetes-deployments-configmap-change/
+
+const ConfigMapAnnotation = "configmap.hash"
+
+func ComputeCurrentConfigMap(ctx context.Context, cli client.Client, cm *corev1.ConfigMap) (string, error) {
+	updatedConfigMap := &corev1.ConfigMap{}
+	key := client.ObjectKeyFromObject(cm)
+
+	if err := cli.Get(ctx, key, updatedConfigMap); err != nil {
+		// ConfigMap not created yet, use the data from the manifests
+		if apierrors.IsNotFound(err) {
+			updatedConfigMap = cm
+		} else {
+			return "", errors.Wrapf(err, "could not calculate ConfigMap %q hash", key.String())
+		}
+	}
+	cmHash := ConfigMapData(updatedConfigMap)
+	klog.InfoS("configmap hash calculated", "hash", cmHash)
+	return cmHash, nil
+}
+
+func ConfigMapData(cm *corev1.ConfigMap) string {
+	var dataAsString string
+
+	if cm.Data != nil {
+		dataAsString = fmt.Sprintf("%v", cm.Data)
```

Comment:
I'm not sure this representation is stable, meaning that we always get the same string across different golang versions. But let's address in a followup PR

Reply:
hmm, I'm open to suggestions in a follow-up PR.
Besides manually serializing the data I don't see any other easy option here.

Reply:
for similar needs, kube uses `spew`, which was later obsoleted and replaced by https://pkg.go.dev/github.com/kr/pretty
see: https://github.com/kubernetes/kubernetes/pull/102882/commits/2f426fdba6fa0458c4a47dd22b37cc920d9f2a98#diff-3b7b6a45fe39aa8a907c6a88b4f067e54c82461ac12da73c0c92d678274b5493R50

The golang formatting per se is as much as stable as third pary packages, but we need to make sure we traverse maps always in the same ordering (= sort keys), which is solved for free by packages like `pretty` (but let's doublecheck)